src/coreclr/vm/excep.cpp

```diff
@@ -6533,59 +6533,30 @@ AdjustContextForJITHelpers(
 
 #if defined(USE_FEF) && !defined(TARGET_UNIX)
 
-struct SavedExceptionInfo
+struct HandleManagedFaultParam
 {
-    EXCEPTION_RECORD m_ExceptionRecord;
-    CONTEXT m_ExceptionContext;
-    CrstStatic m_Crst;
-
-    void SaveExceptionRecord(EXCEPTION_RECORD *pExceptionRecord)
-    {
-        LIMITED_METHOD_CONTRACT;
-        size_t erSize = offsetof(EXCEPTION_RECORD, ExceptionInformation) +
-            pExceptionRecord->NumberParameters * sizeof(pExceptionRecord->ExceptionInformation[0]);
-        memcpy(&m_ExceptionRecord, pExceptionRecord, erSize);
-
-    }
-
-    void SaveContext(CONTEXT *pContext)
-    {
-        LIMITED_METHOD_CONTRACT;
-#ifdef CONTEXT_EXTENDED_REGISTERS
-
-        size_t contextSize = offsetof(CONTEXT, ExtendedRegisters);
-        if ((pContext->ContextFlags & CONTEXT_EXTENDED_REGISTERS) == CONTEXT_EXTENDED_REGISTERS)
-            contextSize += sizeof(pContext->ExtendedRegisters);
-        memcpy(&m_ExceptionContext, pContext, contextSize);
+    EXCEPTION_RECORD *pExceptionRecord;
+    CONTEXT *pContext;
+};
 
-#else // !CONTEXT_EXTENDED_REGISTERS
+LONG FixContextForFaultingExceptionFrame(EXCEPTION_POINTERS* ep, LPVOID pv)
+{
+    WRAPPER_NO_CONTRACT;
 
-        size_t contextSize = sizeof(CONTEXT);
-        memcpy(&m_ExceptionContext, pContext, contextSize);
+    HandleManagedFaultParam *pParam = (HandleManagedFaultParam *) pv;
 
-#endif // !CONTEXT_EXTENDED_REGISTERS
-    }
+    // don't copy param args as have already supplied them on the throw
+    memcpy((void*) ep->ExceptionRecord,
+           (void*) pParam->pExceptionRecord,
+           offsetof(EXCEPTION_RECORD, ExceptionInformation)
+          );
 
-    DEBUG_NOINLINE void Enter()
-    {
-        WRAPPER_NO_CONTRACT;
-        ANNOTATION_SPECIAL_HOLDER_CALLER_NEEDS_DYNAMIC_CONTRACT;
-        m_Crst.Enter();
-    }
+    ReplaceExceptionContextRecord(ep->ContextRecord, pParam->pContext);
```

Comment:
does this handle the special casing of non-catchable exceptions?

Reply:
We need to restore the original context record for diagnostic purposes.  Moreover, handling of the `STATUS_ACCESS_VIOLATION` code depends on having the correct IP address in the context record.  If the IP address belongs to managed code and the failing address is in the first 64 KiB, we translate that to a harmless `NullReferenceException`.

Reply:
I restored this snippet from #70428.  We need to restore the original context record for diagnostic purposes.  Moreover, handling of the `STATUS_ACCESS_VIOLATION` code depends on having the correct IP address in the exception record.  If the IP address belongs to managed code and the address we failed to access is in the first 64 KiB, we translate that to a harmless `NullReferenceException`.